### ISSUE TITLE
Training data and model codec alphabets mismatch

### ISSUE THREAD

**Issue by SarimAshfaq** [mittagessen/kraken#195]
After training my model I tried to fine tune it with additional dataset and it gave me following error
 `[6.8689] Training data and model codec alphabets mismatch: {'۳', '۰', 'E', 'ٰ', 'ٓ', '۲'} `
the command I uses is `ketos train -i model_best.mlmodel syr/*.png`, and "syr" folder contains .png, .gt.txt and manifest flies.
I'm training this model for Urdu language.

**Comment by dstoekl** [comment]
you need to add --resize add

**Comment by SarimAshfaq** [comment]
> you need to add --resize add

@dstoekl adding --resize add worked if I train my model on images I trained it initially on. But when I switch to new (additional) images, it gives following warnings 
`WARNING:kraken:alphabet mismatch: chars in training set only: {'چ', 'ٹ', 'ث', '۳', '؟', 'غ', 'ض', 'ئ', 'ذ', 'ط', 'E', 'ً', ')', '5', 'ڑ', "'", '!', 'آ'} (
not included in accuracy test during training)`
`WARNING:kraken:alphabet mismatch: chars in validation set only: {'ٰ', '۰', '۲'} (not trained)
Network codec not compatible with training set`

and error
`ERROR:kraken:Training data and model codec alphabets mismatch: {'ٓ', 'E', '۳'}`

**Comment by dstoekl** [comment]
each time your training ground truth contains at least one character not included in the previous run you need to add this flag again.

**Comment by SarimAshfaq** [comment]
@dstoekl thanks it worked, But the model's accuracy is relatively low now. Even evaluation of my model on initial images gave lower accuracy than before. Could it be because of images I'm using?

**Comment by dstoekl** [comment]
Difficult without seeing the material or having a more detailed description. How many images do you use with how many characters on them on the average? How many different glyphs do you have? Is it one font or several? BW or graycale? Resolution?

**Comment by mittagessen** [comment]
* Sarim Ashfaq :: 2020-04-28 14:32 Tue:
> @dstoekl thanks it worked, But the model's accuracy is relatively low
> now. Even evaluation of my model on initial images gave lower accuracy
> than before. Could it be because of images I'm using? 

It is expected that the initial accuracy will be lower as the network is
basically brute forced to accomodate the additional characters, making
all the weight slightly wrong. It is also possible that the second stage
training data causes the network to unlearn what is learnt in the
initial training. Lastly, there is a possibility that there's an error
in the 2nd stage training data, e.g. the images are grayscale while the
original ones were black and white which can be caused inadvertently by
conversion into JPG (which can't contain B/W data). The development
version of kraken now warns when one attempts to mix data in this
fashion.

How many lines are we talking about?
